### PR TITLE
Scenario vanilla oversights patch system

### PR DESCRIPTION
**Pull request by tupaschoal**
On #19657 @ZehMatt and I discussed about a "patch system" that would carry all fixes for vanilla scenarios, so that we could clean up code by doing a single generic function to apply the patch. This is me deciding to give it a go, of course this **code is not final**, but it **works**, and that was my goal, get something up and the discussion going.

Some things to think about, in case we want to move forward:
1. If we want to have some sort of JSON schema to validate against on new PRs. There's no built-in function for this on nlohmann::json, so likely it would have to be done on command line
2. If the patch will have to have all sections, even if empty, or if we allow to just provide the ones needed.
3. If we will enforce some/any json formatting
4. If the patches will be on a separate repository
5. Etc...

### CODE REVIEW

**Comment by ZehMatt** [comment]
I think the patch filename should be the md5 sum of the park/sv4/sv6 file, if we store it in the json then we have to load all patches first, if the name matches something like <md5>.parkpatch then we just need to check if it exists. But glad that you picked it up, quite busy to do anything big lately.

**Comment by tupaschoal** [comment]
Ah, but I was not hoping to store it in the JSON, I was thinking about the filename matching the scenario, but with a different extension. MD5 would work too, but I think it will be very ugly for little gain

**Comment by ocalhoun6** [comment]
But some people might be using different versions of the park files, but with the same filename. Especially for RCT1 parks, because there are a bunch of RCT1 scenario recreations for RCT2 out there that people might be using ... and those might be using exactly the same filename as the original RCT1 park ... but it could have important differences, such as already fixing some of the missing tiles, etc.

**Comment by tupaschoal** [comment]
That's a good point, there's an easy way out to limit the patching to just the original formats or read ignoring the extension but then seeing if a hash matches.

**Comment by tupaschoal** [comment]
As of 81f42e0 I have completed the land ownership patch system for all types needed, testing on parks that required at least one of the types and also supporting the `doNotDowngrade` parameter.

Next step is to convert all remaining land ownership fixes to JSON.

One sad thing is that JSON formatters leave the JSON very long, I'd rather have the formatting we already have on our importers, but on the other hand not applying a formatter will make them inconsistent (not sure if it matters much).

After I convert the land ownership, I'll need a small patch to support `fixWater()`. Not sure yet whether I'll convert `FixAyersRockScenario` on `S6Importer` or `FixUrbanPark` on `S4Importer`, as they seem overly specific (but probably worth it)

**Comment by ZeeMaji** [comment]
A more modular FixUrban park would be useful for #20095 as this scenario also has the same issue with the swapped entrance/exit huts. Would also be useful for fixing the RCTC version of urban park assuming the same patch file can't be used for both the RCT1 & RCTC versions of the scenario.

**Comment by tupaschoal** [comment]
I have split into:
- 1c904ba - Infrastructure for reading land ownership patches
- 3fc0ca8 - Conversion of all S6 patches into JSON files and cleanup of S6Importer

Still pending:
- Cleanup S4Importer
- fixWater support
- Specific scenarios support
- Some hardcoded tests that iterate over all patches to make sure they load?
- md5 naming/checking

**Comment by ZeeMaji** [comment]
> But some people might be using different versions of the park files, but with the same filename. Especially for RCT1 parks, because there are a bunch of RCT1 scenario recreations for RCT2 out there that people might be using ... and those might be using exactly the same filename as the original RCT1 park ... but it could have important differences, such as already fixing some of the missing tiles, etc.

For what it's worth, as far as i know there are only 3 official versions of RCT1 scenarios. With only 2 of them being relevent enough to matter

1. The original .sc4 files provided in RCT1 and it's expansions
2. The "modernized" .sea files provided in RCT Classic
3. The few "Classic RollerCoaster Tycoon" sc6 scenarios released as official DLC for rct2 back in the day (https://rct2.rctspace.com/scenarios.asp)

Imo No. 3 can be safely ignored as these versions of the scenarios are very old and obscure (not to mention very incomplete.) So if you were to only apply fixes to .sc4 and .sea you would still cover all of the relevent official versions of the scenarios and automatically ignore all fan recreations which are all sc6.

**Comment by ocalhoun6** [comment]
Can you really be sure that no fan-made .sc4 versions exist, though?

Or possibly .sea files. If someone with RCT Classic made a set of scenarios to be imported into RCT2, they would probably be .sea.

**Comment by ZeeMaji** [comment]
Fan-made sc4s exist, but they primarily use scenario IDs of the competition/unused scenarios (which are in the hundreds afaik) so that shouldn't cause any issues as none of the competiton parks have any patches afaik, However assuming you do this by file name and not scenario ID then that eliminates most custom sc4 scenarios as they either already have unique file names or will need to be renamed anyways so as not to conflict with the original competition scenarios.

I was generally assuming for s4 this would be a check for the internal scenario ID which i believe is used for the current fix code (Eg: Forest frontiers is id 0)
Or the actual full file name with extension (Eg: Forest frontiers is sc0.sc4)

Full file names w/ extension would be less likely to conflict with any custom sc4s due to the limited nature of using IDs within the files themselves, but that would be the path of least conflicts so there's that.

Fan-made .seas do not exist as they are an encrypted import only format that is essentially just an encrypted sc6, RCT Classic does not allow you to create your own .sea files nor are there any tools to create them.

**Comment by tupaschoal** [comment]
Done: 
- fixWater support
- Ayers Rock patch moved into support system

Still pending:
- Urban park move into patch
- Some hardcoded tests that iterate over all patches to make sure they load?
- md5 naming/checking
- Make ´FixLandOwnershipTilesWithOwnership` not take an `initializer_list` and instead take a vector?

Some TODOs found during conversion that might be dropped depending on the pending issues above:
- investigate why Katie's Dreamland.s4 and .sea have different patches
- s4 patch system is broken since .s4 files do not use scenario name on file name, but rather slot
- Check if path is case sensitive, some scenario names have all lowercase variations
- Land ownership is applied even when loading saved scenario. Should it?

**Comment by tupaschoal** [comment]
All fixes are now into patches, now for the gritty validation details.

**Done**: 
- Urban park move into patch

**Still pending**:
- Some hardcoded tests that iterate over all patches to make sure they load?
- md5 naming/checking
- Make `FixLandOwnershipTilesWithOwnership` not take an `initializer_list` and instead take a vector?
- Move extension from `.json` to something like `.parkpatch` as suggested by Matt
- Remove `cout`

Some TODOs found during conversion that might be dropped depending on the pending issues above:
- investigate why Katie's Dreamland.s4 and .sea have different patches
- s4 patch system is broken since .s4 files do not use scenario name on file name, but rather slot
- Check if path is case sensitive, some scenario names have all lowercase variations
- Land ownership is applied even when loading saved scenario. Should it?

**Comment by HtotheTML** [comment]
I'm uncertain about the logistics of applying fixes when loading a scenario; I would hope loading a saved file doesn't reapply the patch, especially when it comes to things like the fixwater patch for botany breakers, which adds a pool of water inside park boundaries. I'd say patches should only apply when loading the scenario for the first time, but not when loading a save, if possible.

**Comment by tupaschoal** [comment]
> I'm uncertain about the logistics of applying fixes when loading a scenario; I would hope loading a saved file doesn't reapply the patch, especially when it comes to things like the fixwater patch for botany breakers, which adds a pool of water inside park boundaries. I'd say patches should only apply when loading the scenario for the first time, but not when loading a save, if possible.

That's already what happens with most of the fixes, except for the land ownership (which I think might have been an oversight)

**Comment by karst** [comment]
Wouldn't it be smart, once you're done with this, to move all patches to #OpenScenarios? Create a release for that and download them here?

**Comment by tupaschoal** [comment]
> Wouldn't it be smart, once you're done with this, to move all patches to #OpenScenarios? Create a release for that and download them here?

Not sure if OpenScenarios is the best thing, but yes, I thought about it, so we don't keep these objects around this repo, just haven't thought through because want to get the issues done first.

**Comment by tupaschoal** [comment]
**Done**:
- Make FixLandOwnershipTilesWithOwnership not take an initializer_list and instead take a vector
- Unify method to read coordinates


**Still pending**:
- Some hardcoded tests that iterate over all patches to make sure they load?
- md5 naming/checking
- Move extension from .json to something like .parkpatch as suggested by Matt
- Remove cout

Some TODOs found during conversion that might be dropped depending on the pending issues above:
- investigate why Katie's Dreamland.s4 and .sea have different patches
- s4 patch system is broken since .s4 files do not use scenario name on file name, but rather slot
- Check if path is case sensitive, some scenario names have all lowercase variations
- Land ownership is applied even when loading saved scenario. Should it?

**Comment by karst** [comment]
Reason I was suggesting OpenScenarios was because how things are done in OpenMusic and OpenSoundEffects. Then again, default music objects are in the objects repo. I could see them fit in either.

**Comment by tupaschoal** [comment]
**Done**:

- Rebased to account for:
  - #21179
  - #21309
  - #21316

**Still pending**:

- Some hardcoded tests that iterate over all patches to make sure they load?
- md5 naming/checking
- Move extension from .json to something like .parkpatch as suggested by Matt
- Remove cout

Some TODOs found during conversion that might be dropped depending on the pending issues above:
- investigate why Katie's Dreamland.s4 and .sea have different patches
- s4 patch system is broken since .s4 files do not use scenario name on file name, but rather slot
- Check if path is case sensitive, some scenario names have all lowercase variations
- Land ownership is applied even when loading saved scenario. Should it?

**Comment by tupaschoal** [comment]
**Done**:

- Rebased to account for: #21425
- SHA256 naming/checking: Every parkpatch file has the full SHA256 of the scenario, but its filename is a shortened version just for readability. When loading, the code does proper checking to ensure it didn't mistakenly load a conflicting hash
- Move extension from .json to something like .parkpatch as suggested by Matt
- Not using filename anymore, so this stopped being a concern: 
  - ~s4 patch system is broken since .s4 files do not use scenario name on file name, but rather slot~ 
  - ~Check if path is case sensitive, some scenario names have all lowercase variations~ 
 
**Still pending**:
- Some hardcoded tests that iterate over all patches to make sure they load?
- Replace cout with LOGGING messages
- Calculate hashes of `.sea` files and duplicate relevant `.parkpatch`es

Some TODOs found during conversion that might be dropped depending on the pending issues above:
- investigate why Katie's Dreamland.s4 and .sea have different patches
- Land ownership is applied even when loading saved scenario. Should it?

**Comment by tupaschoal** [comment]
**Done**:
- Hardcoded tests that iterate over all patches to make sure they load
- Replace cout with LOGGING messages
- Calculate hashes of `.sea` files and duplicate relevant `.parkpatch`es
- investigate why Katie's Dreamland.s4 and .sea have different patches

**TODOs to carry forward after PR is merged**:
- Land ownership is applied even when loading saved scenario. Should it?
- Seems like `.sea` files are loaded twice (saw patch algorithm running twice in a row)
- RCTC scenarios (`.sea`) are a collection of scenarios from RCT1, RCT2 and their expansions. They seem to be slightly different, such that some bugs do not exist on the `.sea` (like several of Katie's Dreamland patches). Maybe we should review all `.sc4` and `.sc6` that have patch files to make sure they are also needed on `.sea`. Also it would be nice to see whether some of the `.sea` patches are irrelevant.

**Comment by AaronVanGeffen** [comment]
@tupaschoal What's the status of this? Do you need input on anything in particular?

**Comment by AaronVanGeffen** [comment]
> It was mostly ready, but then I hit the issue that for builds with no network, there is no cipher algorithm to use. We can bundle a header only implementation of it to overcome that, if people agree.

Assuming you need a hash function, what about the [OpenRCT2FNV1aAlgorithm class](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/core/Crypt.OpenRCT2.cpp)? I believe that implements the [Fowler–Noll–Vo hash function](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function).

**Comment by tupaschoal** [comment]
> Assuming you need a hash function, what about the [OpenRCT2FNV1aAlgorithm class](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/core/Crypt.OpenRCT2.cpp?rgh-link-date=2024-07-12T22%3A01%3A47Z)? I believe that implements the [Fowler–Noll–Vo hash function](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function).

I can try that. What I like about SHA256 is that people can even find websites to generate them, so it would make creating new `.parkpatch`es easier, the same is not so true for FNV-1a

**Comment by AaronVanGeffen** [comment]
That's a good point. I'm personally not against getting in a dependency for SHA256 hashing, either, but I worry it won't be tested enough. All release builds shipped by us have the network component enabled, as far as I'm aware.

We could also just don't build this patch component when the network component is not included. I think that would affect a relatively small portion of our user base, if any. These players wouldn't be missing anything, either, as they've never had it in the first place.

**Comment by tupaschoal** [comment]
> These players wouldn't be missing anything, either, as they've never had it in the first place.

That's not entirely true, is it? All the patches are currently hardcoded and work for any build type, so they would lose some functionality, but...

> All release builds shipped by us have the network component enabled, as far as I'm aware.

If we don't ship any non-networked builds, should we even care?

**Comment by tupaschoal** [comment]
@AaronVanGeffen I've disabled the SHA code if network is disabled so now all the patches pass. Also solved merged conflict.
I'm marking the PR as ready for review

**Comment by ZehMatt** [comment]
I'm not sure I'm a fan of coupling this with the network functionality, wouldn't it be better to decouple the hashing so it works regardless on how its built?

**Comment by ZehMatt** [comment]
> > I'm not sure I'm a fan of coupling this with the network functionality, wouldn't it be better to decouple the hashing so it works regardless on how its built?
> 
> Yes, but then there are the considerations I raised a couple messages back

Yeah I get that, I still think its best to just use thirdparty code to get SHA256 out of the box with or without network functionality, having the patch system locked behind the network macro feels pretty wrong.

**Comment by tupaschoal** [comment]
Added one header-only library to have sha256 then

File: src/openrct2/rct12/ScenarioPatcher.cpp
Comment (tupaschoal) [inline src/openrct2/rct12/ScenarioPatcher.cpp:34]:
```suggestion
#ifdef DISABLE_NETWORK
#    include <picosha2.hpp>
#else
#    include "../core/Crypt.h"
#endif
```

**Comment by AaronVanGeffen** [comment]
@ZehMatt Are you satisfied with the new situation?

**Review by ZehMatt — APPROVED** [review]
I think this is pretty good, we can definitely add more features in the future and perhaps even automate the generation of patches at some point, at least now we wont need to add code to make up for content fixes.

**Comment by tupaschoal** [comment]
@HtotheTML since you contributed with a lot of fixes, do you want to take this up for a spin?

**Comment by HtotheTML** [comment]
Sure! I'm a bit hazy on the process of testing another fork, so I may need a reminder in the Discord.

**Comment by HtotheTML** [comment]
Cross referencing the modifications with previous PRs that made these fixes, I noticed the following errors:

Alton Towers' changes became buyable land rather than owned land
Alton Towers (11, 31)
(11, 31)

Alton Towers (68, 112) (72, 118)
(68, 112) (72, 118)

Aqua Park's changes were turned into owned construction rights rather than owned land
Aqua Park (42, 85)
(42, 85)

Aqua Park (89, 42)
(89, 42)

Schneider Shores contains two new spots that were overlooked for corrections previously; these need to be buyable land rather than buyable construction rights
Schneider Shores (94-95,70)
(94-95, 70)

Schneider Shores (137, 80)
(137, 80)

All of the errors in Six Flags Magic Mountain were reverted to their unpatched state. These should all be buyable land (I bought all the available land in-game to make it easier to find a number of them
Six Flags Magic Mountain (46, 51)
(46, 51)

Six Flags Magic Mountain (58, 159)
(58, 159)

Six Flags Magic Mountain (61, 92-95) (62, 90-94)
(61, 92-95) (62, 90-94)

Six Flags Magic Mountain (71, 201)
(71, 201)

Six Flags Magic Mountain (75, 167)
(75, 167)

Six Flags Magic Mountain (89, 40-42) (88, 42)
(88, 42) (89, 40-42)

Six Flags Magic Mountain (92-93, 57)
(92-93, 57)

Six Flags Magic Mountain (104-105, 190) (108, 97)
(104-105, 190) (108, 97)

Six Flags Magic Mountain (126, 15)
(126, 15)

Six Flags Magic Mountain (168-169, 20)
(168-169, 20)

Six Flags Magic Mountain (190, 6)
(190, 6)

A number of scattered tiles in Urban Park are now owned land rather than buyable land
Urban Park (39, 20)
(39, 20)

Urban Park (64, 77) (61, 66-67)
(61, 66-67) (64, 77)

Additionally, many tiles are both purchasable land rights and purchasable land that should only be purchasable land. I have circled the tiles this applies to, since it's difficult to show with only one image
Urban Park (34, 60-61) (35, 60-61)
(34, 60-61) (35, 60-61)

Urban Park (48, 19-20) (52-55, 15) (52,55, 16) (52, 17-19)
(48, 19-20) (52-55, 15) (52-55, 16) (52, 17-19)

All of the other boundary fixes appear to be in working order.

**Comment by tupaschoal** [comment]
This is amazing @HtotheTML , thank you very much!
I'll get to it ASAP

**Comment by HtotheTML** [comment]
Did plans exist for incorporating water error fixes into this patch system? Like the one in Botany Breakers.

**Comment by tupaschoal** [comment]
> Did plans exist for incorporating water error fixes into this patch system? Like the one in Botany Breakers.

We already have such a system, see `fcc15f9.parkpatch` (Infernal Views) for an example.


I have applied all the fixes you pointed out, including the new ones. Also double checked the migration on my own and updated the names that had been updated since this PR was first created. This is now ready to merge again.

**Comment by HtotheTML** [comment]
Am I missing something, then? I pulled up the ones mentioned in #19740 with this build and they seem to be left unchanged...

**Comment by tupaschoal** [comment]
That PR was not merged, but if you pull not sure how you pulled the changes, bug we can add those fixes to the JSON after this one is merged .

**Comment by HtotheTML** [comment]
Ooooohh gotcha. No worries, then! Apologies for the confusion.

**Comment by Gymnasiast** [comment]
I took a broad look at the code and it looks structurally fine. It looks like it must have been a lot of work!

I’ll take a closer look/test soon.

**Comment by Gymnasiast** [comment]
> Land ownership is applied even when loading saved scenario. Should it?

This was a deliberate choice on my part. It’s also why I added the "cannot downgrade" flag, to minimise the chance of the fixes mucking up saves where someone already bought land, or used cheats to own everything. I would be open to limiting all fixes to just scenarios though.

**Comment by tupaschoal** [comment]
I think if we limit it to just scenarios when first loading, we can drop these two and simplify. I don't think it's worth fixing it in this PR, though. I intend to tackle all todos on next PRs

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**
